odxtools/request.py

```diff
@@ -1,39 +1,114 @@
+                    composite_codec_get_required_parameters, composite_codec_get_static_bit_length)
+from .decodestate import DecodeState
+from .element import IdentifiableElement
 from .encodestate import EncodeState
-from .odxlink import OdxDocFragment
-from .odxtypes import ParameterValue
+from .exceptions import odxraise
+from .nameditemlist import NamedItemList
+from .odxlink import OdxDocFragment, OdxLinkDatabase, OdxLinkId
+from .odxtypes import ParameterValue, ParameterValueDict
+from .parameters.createanyparameter import create_any_parameter_from_et
+from .parameters.parameter import Parameter
 from .snrefcontext import SnRefContext
+from .specialdatagroup import SpecialDataGroup
 from .utils import dataclass_fields_asdict
 
 
-# TODO: The spec does not say that requests are basic structures. For
-# now, we derive from it anyway because it simplifies the en- and
-# decoding machinery...
 @dataclass
-class Request(BasicStructure):
+class Request(IdentifiableElement):
+    """Represents all information related to an UDS request
+
+    This class implements the `CompositeCodec` interface.
+    """
+    admin_data: Optional[AdminData]
+    parameters: NamedItemList[Parameter]
+    sdgs: List[SpecialDataGroup]
+
+    @property
+    def required_parameters(self) -> List[Parameter]:
+        return composite_codec_get_required_parameters(self)
+
+    @property
+    def free_parameters(self) -> List[Parameter]:
+        return composite_codec_get_free_parameters(self)
 
     @staticmethod
     def from_et(et_element: ElementTree.Element, doc_frags: List[OdxDocFragment]) -> "Request":
-        """Reads a response."""
-        kwargs = dataclass_fields_asdict(BasicStructure.from_et(et_element, doc_frags))
+        kwargs = dataclass_fields_asdict(IdentifiableElement.from_et(et_element, doc_frags))
 
-        return Request(**kwargs)
+        admin_data = AdminData.from_et(et_element.find("ADMIN-DATA"), doc_frags)
+        parameters = NamedItemList([
+            create_any_parameter_from_et(et_parameter, doc_frags)
+            for et_parameter in et_element.iterfind("PARAMS/PARAM")
+        ])
+        sdgs = [
+            SpecialDataGroup.from_et(sdge, doc_frags) for sdge in et_element.iterfind("SDGS/SDG")
+        ]
 
-    def encode(self, **kwargs: ParameterValue) -> bytes:
-        encode_state = EncodeState(is_end_of_pdu=True)
+        return Request(admin_data=admin_data, parameters=parameters, sdgs=sdgs, **kwargs)
 
-        self.encode_into_pdu(physical_value=kwargs, encode_state=encode_state)
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        result = {self.odx_id: self}
 
-        return encode_state.coded_message
+        for param in self.parameters:
+            result.update(param._build_odxlinks())
+
+        return result
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        for param in self.parameters:
+            param._resolve_odxlinks(odxlinks)
```

Comment:
what about the other properties?

Reply:
I am asking why not call _resolve_odxlinks for the admin data and sdgs?

Reply:
ops. right, fixed. Thanks!